http_parser.c

```diff
@@ -472,6 +483,89 @@ static struct {
 
 int http_message_needs_eof(const http_parser *parser);
 
+const char* find_crlf(const char* p, const char* data, size_t len);
+
+#if defined(_MSC_VER) // SSE2 is baseline
+#include <intrin.h>
+#define __builtin_ctz _tzcnt_u32
+#elif defined(__SSE2__) && defined(__GNUC__)
+#include <emmintrin.h>
+#endif
+
+#ifndef USE_INTRISICS_CRLF
+# define USE_INTRISICS_CRLF defined(_MSC_VER) || (defined(__SSE2__) && defined(__GNUC__))
+#endif
+
+#if USE_INTRISICS_CRLF
+
+int32_t get_crlf_mask(const char* p) {
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  const __m128i vCR = _mm_set1_epi8(0x0a);
+
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  const __m128i vLF = _mm_set1_epi8(0x0d);
+
+  __m128i v1, v2;
+  v1 = *(__m128i*)p;
+  v2 = _mm_cmpeq_epi8(vCR, v1);
+  v1 = _mm_cmpeq_epi8(vLF, v1);
+  v2 = _mm_or_si128(v1, v2);
+  return _mm_movemask_epi8(v2);
+}
+
+const char* find_crlf(const char* p, const char* data, size_t len) {
+  const char* lastp = MIN(data + len, HTTP_MAX_HEADER_SIZE + p);
+
+  int32_t result = 0;
+
+  size_t alignment = (long) p & 15;
+  p -= alignment;
+
+  result = get_crlf_mask(p + 16) << 16 | get_crlf_mask(p);
```

Comment:
Don't we need to check that `len >= 16` before trying to use the intrinsics which operate on 16 bytes at a time? For example, a valid HTTP request could just be `GET /<CR><LF><CR><LF>` which is only 9 bytes.

Reply:
Actually unless there are some safety or memory concerns, then no, the logic is as such

##case1
data: \r\nGET /\r\n\r\n
p: data+2
len: 9
and imagine if alignment is off by 1

*get alignment offset
*p -= 1, which is data+1
*checks next 32 bytes
*result will be that looks something like 1000 0011 1100 1010 1010 1010 1010 1010(it should be right to left, i wrote it left to write)
**anything beyond the 10th bit is for some data that's beyond the range
*shift right and left by alignment offset, so it will end up being 0000 0011 1100 1010 1010 1010 1010 1010
*and so the first bit found would be the first \r from data+2

##case2

data: \r\nGET /\r\n\r\n
p: data+2
len: 4
and imagine if alignment is off by 1

*exactly the same as before
*but the position of the first found \r after removing alignment false positives will be beyond "lastp"
**lastp is either data+len or max header size + p

Reply:
It reads beyond the end of the input, doesn't it?  Or am I misunderstanding your comment?

Reply:
Ah, yes it does, this pr's been so long, i think i've lost myself somewhere. Let me get that fixed, it should be changed to matching one by one till aligned and then matching remaining one by one

Reply:
Made a change to the logic

while misaligned and not last char
  checks char one by one
  returns if found
if remaining is less than 32:
  while not last char
    checks char one by one
    return if found
  return not found
SIMD match till there are at least 32 left, return if found
if not found and not ended yet
  find char one by one and return if found
return if not found

this change will also ensure that we will not be offsetting the pointer to a possibly bogus address, nor will we read beyond the input anymore.